core/systemcontractupgrade/system-contract-upgrade.go

```diff
@@ -0,0 +1,116 @@
+package systemcontractupgrade
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type UpgradeConfig struct {
+	BeforeUpgrade beforeUpgrade
+	AfterUpgrade  afterUpgrade
+	ContractAddr  common.Address
+	CommitUrl     string
+	Code          string
+}
+
+type Upgrade struct {
+	UpgradeName string
+	Configs     []*UpgradeConfig
+}
+
+type beforeUpgrade func(blockNumber *big.Int, contractAddr common.Address, statedb *state.StateDB)
```

Comment:
there is no need to define type `beforeUpgrade ` and `afterUpgrade`, `upgradeHook` is enough.

Reply:
please add `return (error)` to this hook.

Reply:
> there is no need to define type `beforeUpgrade ` and `afterUpgrade`, `upgradeHook` is enough.

I don't think upgradeHook is enough. Sometime we have to do cleanup work before update code, and sometime we have to do init work after update code

Reply:
Sorry for misunderstand. We can just define one function type`upgradeHook` but multiple instances
```
type UpgradeConfig struct {
	BeforeUpgrade ugradeHook
	AfterUpgrade  ugradeHook
        ...
}
```